### PR TITLE
Remove Item::kind, use tagged enum. Rename variants to match

### LINKED ISSUE #82299: Rustdoc Json: structs and functions deserialize incorrect
I tried this code:

```rust
pub struct Foobar {
    field: u8,
}
```

Running `rustdoc` on this this happily produces the following json which I would judge to be correct.

<details>

```
{"root":"0:0","crate_version":null,"includes_private":true,"index":{"0:2":{"id":"0:2","crate_id":0,"name":"field","source":{"filename":"src/main.rs","begin":[4,4],"end":[4,13]},"visibility":"default","docs":null,"links":{},"attrs":[],"deprecation":null,"kind":"struct_field","inner":{"kind":"primitive","inner":"u8"}},"0:1":{"id":"0:1","crate_id":0,"name":"Foobar","source":{"filename":"src/main.rs","begin":[3,0],"end":[5,1]},"visibility":"public","docs":null,"links":{},"attrs":[],"deprecation":null,"kind":"struct","inner":{"struct_type":"plain","generics":{"params":[],"where_predicates":[]},"fields_stripped":false,"fields":["0:2"],"impls":[]}},"0:0":{"id":"0:0","crate_id":0,"name":"rustdoc_issue","source":{"filename":"src/main.rs","begin":[1,0],"end":[5,1]},"visibility":"public","docs":null,"links":{},"attrs":["#![feature(no_core)]","#![no_core]"],"deprecation":null,"kind":"module","inner":{"is_crate":true,"items":["0:1"]}}},"paths":{"0:0":{"crate_id":0,"path":["rustdoc_issue"],"kind":"module"},"0:1":{"crate_id":0,"path":["rustdoc_issue","Foobar"],"kind":"struct"}},"external_crates":{},"format_version":4}
```

</details>

However, deserializing this with the types defined in `src/rustdoc-json-types/lib.rs` leads to the incoherent representation:

```rust
Item {
    kind: ItemKind::Struct,
    inner: ItemEnum::Union { … }
}
```

This happens because the `ItemEnum` is an untagged enum, the `UnionItem` variant appears before the `StructItem` variant and the `struct Union` has a subset of the fields of `struct Struct`, therefore it matches first. The same issue happens with `FunctionItem`/`Function` and `MethodItem`/`Method`.

It could be fixed by reordering the variants but that's quite brittle and non-obvious.

### Meta

`rustc --version --verbose`:
```
 rustc 1.52.0-nightly (a143517d4 2021-02-16)
```

cc: @jyn514

### PR DESCRIPTION
**Pull request by CraftSpider**
Fixes #82299, by making the ItemEnum tagged. Doesn't remove ItemKind as it's still used in other places.

r? @jyn514 
@rustbot label: +A-rustdoc-json +T-rustdoc

### CODE REVIEW

**Comment by CraftSpider** [comment]
CC @HeroicKatora

**Comment by aDotInTheVoid** [comment]
Can you add a test for this, (probably in `rustdoc-json-types`, that round trip serilization and deserialization keeps it the same)

Also nit: To be more consistant with `clean::types` the `inner` field should be renamed to `kind`, and the `ItemEnum` should be renamed to `ItemKind` (see [here](https://doc.rust-lang.org/nightly/nightly-rustc/rustdoc/clean/types/struct.Item.html))

**Comment by CraftSpider** [comment]
I'll add a test. I'm not sure I like that naming, I think 'kind' is a weird name for the inner values, and we don't really need consistency with clean when it's not harming functionality. Also, `ItemKind` already exists in JSON, so that name can't be used again.

**Comment by jyn514** [comment]
> I'll add a test. I'm not sure I like that naming, I think 'kind' is a weird name for the inner values, and we don't really need consistency with clean when it's not harming functionality. Also, `ItemKind` already exists in JSON, so that name can't be used again.

See https://github.com/rust-lang/rust/issues/80142, I don't think this is worth changing.

File: Cargo.lock [hunk 1]
```diff
@@ -1,5 +1,7 @@
 # This file is automatically @generated by Cargo.
 # It is not intended for manual editing.
+version = 3
+
```
Comment (aDotInTheVoid) [inline Cargo.lock:4]: Why was this changed. Did you build with nightly cargo.

File: src/librustdoc/json/mod.rs [hunk 2]
```diff
@@ -244,7 +243,7 @@ impl<'tcx> FormatRenderer<'tcx> for JsonRenderer<'tcx> {
                     )
                 })
                 .collect(),
-            format_version: 4,
+            format_version: 5,
```
Comment (aDotInTheVoid) [inline src/librustdoc/json/mod.rs:246]: I dont think we need to change this, as their are no changes to the output, right?

File: src/rustdoc-json-types/tests.rs [hunk 3]
```diff
@@ -0,0 +1,34 @@
+use super::*;
+
+#[test]
+fn test_struct_info_roundtrip() {
```
Comment (aDotInTheVoid) [inline src/rustdoc-json-types/tests.rs:4]: I can't run these. I think you need to do some things in `src/bootstrap` so it knows to run `cargo test` in ` src/rustdoc-json-types`. (Doing this my hand might have caused the `version` in `Cargo.toml`)

```
$ ./x.py test src/rustdoc-json-types
Updating only changed submodules
Submodules updated in 0.02 seconds
    Finished dev [unoptimized + debuginfo] target(s) in 2.26s
thread 'main' panicked at 'error: no rules matched src/rustdoc-json-types', src/bootstrap/builder.rs:243:17
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
failed to run: /home/nixon/upstreams/rust/rust/build/bootstrap/debug/bootstrap test src/rustdoc-json-types
Build completed unsuccessfully in 0:00:03
```

**Review by aDotInTheVoid — CHANGES_REQUESTED** [review]
Mostly good, just a couple of things to do. Also you need to rebase on master

@rustbot modify labels: -S-waiting-on-review +S-waiting-on-author

File: src/rustdoc-json-types/tests.rs (hunk 3 above)
Comment (jyn514) [reply src/rustdoc-json-types/tests.rs:4]: I think you could add it as a `host_test!`: https://github.com/rust-lang/rust/blob/51748a8fc77283914d4135f31b5966a407208187/src/bootstrap/test.rs#L977

Comment (CraftSpider) [reply src/rustdoc-json-types/tests.rs:4]: I think running by hand changed it, yeah. I'll revert the version = 3 change, and fix up bootstrap.

File: src/librustdoc/json/mod.rs (hunk 2 above)
Comment (CraftSpider) [reply src/librustdoc/json/mod.rs:246]: Right, I'll change it back. For some reason I thought I'd changed something, but you're right

File: Cargo.lock (hunk 1 above)
Comment (CraftSpider) [reply Cargo.lock]: I did, it fixed itself when I added the bootstrap test and ran it properly.

File: src/rustdoc-json-types/tests.rs (hunk 3 above)
Comment (CraftSpider) [reply src/rustdoc-json-types/tests.rs:4]: `x.py test src/rustdoc-json-types` now runs the tests as expected
